### DISCUSSION TITLE
Setting up and running Adapter service

### DISCUSSION THREAD

**Discussion by GauravTekdi** [samagra-comms/community#5] [Developer Setup and Troubleshooting]
Hello Team,

We have cloned the fork of **[Adapter repository](https://github.com/samagra-comms/adapter)** and we are using **Intellij**. We have also explored the link shared for **[Adapter.md](https://github.com/samagra-comms/community/blob/main/docs/Adapter.md)** by which we got theoretical basic understanding of **Adapters usage that it is used as a gateway to convert from provider messages to xMesasages and vise-versa.** 
However, we are struggling in setup and running the Adapter service on local. It's giving cannot resolve dependency issues. **Can someone please guide/help us in setup, resolving the dependencies and running the service and understanding the flow/application-structure followed?** 

@aayushkedia95 
cc - @coolbung

**Comment by aayushkedia95** [comment]
@surabhi-mahawar, can you help @GauravTekdi?

cc - @ChakshuGautam

**Comment by surabhi-mahawar** [comment]
Hello @GauravTekdi 

We do not run the adapter service as we are using it as a dependency for inbound & outbound services. What you can do is create your adapter & create tests for it as given in this [example](https://github.com/samagra-comms/adapter/blob/release-4.9.0/src/test/java/com/uci/adapter/netcore/whatsapp/NetcoreServiceTest.java), [example2](https://github.com/samagra-comms/adapter/blob/release-4.9.0/src/test/java/com/uci/adapter/netcore/whatsapp/NetcoreWhatsappAdapterTest.java). 

Please give it a try, If you still face issues, we can have a call around 2 for this today.

cc: @aayushkedia95 @ChakshuGautam

**Reply by GauravTekdi** [reply]
@surabhi-mahawar  Thanks for the response. Let me explore and try these examples you shared.
cc - @coolbung

**Reply by GauravTekdi** [reply]
@surabhi-mahawar Can we connect?

**Reply by surabhi-mahawar** [reply]
@GauravTekdi 
we connect now.
Please join here 
https://meet.google.com/bxs-anaw-irz

**Reply by GauravTekdi** [reply]
@surabhi-mahawar 
We are done with the setup of Adapter on local and successfully resolved the dependency issues over intellij -
- "error:java: jdk isn't specified for module" 
- "java.lang.NoClassDefFoundError: com.intellij.DynamicBundle" and
- "Package 'java.lang' is declared in module 'java.base', which is not in the module graph" errors

And we were able to run the NetcoreWhatsappAdapterTest convertMessageToXMsgTest() successfully, however failing for processOutBoundMessageTest() as it's calling '/message' and giving failing error as "org.springframework.web.reactive.function.client.WebClientResponseException$Forbidden: 403 Forbidden from POST https://waapi.pepipost.com/api/v2/message/". Can you please help us in this?
Also till that have started analyzing for the Email adapter, so can someone please suggest/guide us in the provider request/response format or email format/pattern/fields for inbound and outbound methods? Please guide us for the next step.

**Reply by surabhi-mahawar** [reply]
@GauravTekdi 
The forbidden error you are facing is because the Netcore token you are using is not be allowed to access the netcore outbound apis. To resolve this you need to contact the netcore team to get a token for the apis. we have also mentioned the same in the code.

/** Add NETCORE_WHATSAPP_AUTH_TOKEN in Env file,
 * as we need this to send messaes to netcore, and recieve its response
 */
 
But if you are able to process till this step, means the test it working fine, we are able to test it out till end, as we have a authenticated token from netcore.

For the email adapter you would need a service provider that will send the email message from your system to our inbound apis. The format should be something the service provider will send, like netcore sends us. The example for the sample is given [here](https://github.com/samagra-comms/adapter/blob/release-4.9.0/src/test/java/com/uci/adapter/netcore/whatsapp/NetcoreWhatsappAdapterTest.java#L69).
You will have to write the code to convert this format to XMessage as we are doing in all adapters. Eg. [Netcore incoming message from user api](https://wadocs.pepipost.com/webhooks/overview/incoming-message) 

In inbound we use the adapters to get incoming messages and convert them to XMessage. Here is a simple example of netcore incoming message [Link](https://github.com/samagra-comms/inbound/blob/release-4.9.0/src/main/java/com/uci/inbound/netcore/NetcoreWhatsappConverter.java). You only need to create a adapter for email message, we will then create a inbound url for incoming email messages.

For the outbound process, you will have to convert [XMessage](https://github.com/samagra-comms/adapter/blob/release-4.9.0/src/test/java/com/uci/adapter/netcore/whatsapp/NetcoreWhatsappAdapterTest.java#L87) to a format that your service provider allows, and then send this converted message to your service provider's send message api. Eg. [Netcore send text to user api](https://wadocs.pepipost.com/sending-message/send-message-api/receive-message-api) 

CC: @aayushkedia95 @ChakshuGautam

**Reply by GauravTekdi** [reply]
@surabhi-mahawar 
We have started implementation for the MailgunEMailAdapter similarly to the WhatsappAdapters, however I got up with the some webhook questions like-
- We are implementing the 2 methods, _**convertMessageToXMsg(obj msg), processOutboundMessageF(XMeassage nextMsg)**_, so when does actually these methods would get called? Have we integrated the Netcore or did we registered these implemented methods over the Netcore webhooks to get called back when the certain events occurs?
- Which webhook event callbacks we should use?

cc - @coolbung

**Reply by surabhi-mahawar** [reply]
Hey @GauravTekdi 

We have created an api which is being called by netcore when a user sends message. This api is created in [Inbound](https://github.com/samagra-comms/inbound/blob/release-4.9.0/src/main/java/com/uci/inbound/netcore/NetcoreWhatsappConverter.java). Once your adapter is completed we can create a new api in the same way. The service provider that you will use will target this api. The api will created in a specific format **adapter/channel** (Eg. netcore/whatsapp, gupshup/whatsapp). 

For Outbound there should be an outbound api which will be provided by your service provider. In **processOutboundMessageF** you just have to convert the XMessage to the format which the api accepts and call this api to send your message to api & return the XMessage with sent message [Example](https://github.com/samagra-comms/adapter/blob/release-4.9.0/src/main/java/com/uci/adapter/netcore/whatsapp/NetcoreWhatsappAdapter.java#L386).

CC: @aayushkedia95 @ChakshuGautam
